lib/Mojo/IOLoop/Stream/HTTPServer.pm

```diff
@@ -0,0 +1,246 @@
+        if ($self->{keep_alive} || 1) >= $self->max_requests || $req->error;
+
+      $tx->on(resume => sub { $self->_write_content });
+      $self->_write_content;
+    }
+
+  );
+
+  $self->emit(build_tx => $tx);
+
+  # Kept alive if we have more than one request on the connection
+  return ++$self->{keep_alive} > 1 ? $tx->kept_alive(1) : $tx;
+}
+
+sub _finish {
+  my $self = shift;
+
+  # Always remove connection for WebSockets
+  return unless my $tx = $self->{tx};
+
+  # Finish transaction
+  delete($self->{tx})->closed;
+
+  # Upgrade connection to WebSocket
+  if (my $ws = delete $self->{next}) {
+
+    # Successful upgrade
+    if ($ws->handshake->res->code == 101) {
+      $self->emit(upgrade => $ws->established(1));
+    }
+
+    # Failed upgrade
+    else { $ws->closed }
+  }
+
+  # Close connection if necessary
+  return $self->close_gracefully if $tx->error || !$tx->keep_alive;
+
+  # Build new transaction for leftovers
+  return unless length(my $leftovers = $tx->req->content->leftovers);
+  $self->{tx} = $tx = $self->_build_tx;
+  $tx->server_read($leftovers);
+}
+
+sub _read_content {
+  my ($self, $chunk) = @_;
+  my $tx = $self->{tx} ||= $self->_build_tx;
+  warn term_escape "-- Server <<< Client (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  $tx->server_read($chunk);
+}
+
+sub _url { shift->req->url->to_abs }
+
+sub _write_content {
+  my $self = shift;
+
+  # Protect from resume event recursion
+  return if !(my $tx = $self->{tx}) || $self->{cont_writing};
+  local $self->{cont_writing} = 1;
+  my $chunk = $tx->server_write;
+  warn term_escape "-- Server >>> Client (@{[_url($tx)]})\n$chunk\n" if DEBUG;
+  my $next
+    = $tx->is_finished ? '_finish' : length $chunk ? '_write_content' : undef;
+  return $self->write($chunk) unless $next;
+  $self->write($chunk => sub { shift->$next() });
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojo::IOLoop::Stream::HTTPServer - Non-blocking I/O HTTP server
+
+=head1 SYNOPSIS
+
+  use Mojo::IOLoop::Server;
+  use Mojo::IOLoop::Stream::HTTPServer;
+  
+  # Create listen socket
+  my $server = Mojo::IOLoop::Server->new;
+  $server->on(
+    accept => sub {
+      my $stream = Mojo::IOLoop::Stream::HTTPServer->new(pop);
+  
+      $stream->on(
+        request => sub {
+          my ($stream, $tx) = @_;
+          $tx->res->code(200);
+          $tx->res->headers->content_type('text/plain');
+          $tx->res->body('Hello World!');
+          $tx->resume;
+        }
+      );
+      $stream->start;
+    }
+  );
+  $server->listen(port => 3000);
+  
+  # Start accepting connections
+  $server->start;
+
+=head1 DESCRIPTION
+
+HTTP server connection based on L<Mojo::IOLoop::Stream>.
+
+=head1 EVENTS
+
+L<Mojo::IOLoop::Stream::HTTPServer> inherits all events from
+L<Mojo::IOLoop::Stream> and can emit the following new ones.
+
+=head2 build_tx
+
+  $stream->on(build_tx => sub {
+    my ($stream, $tx) = @_;
+    ...
+  });
+
+Emitted when a transaction for new request was build.
```

Comment:
Can't help but think that this event needs a better name, will have to think about it a little more. (Oh, and of course a better description 😉)